core/src/hash/hasher.rs

```diff
@@ -0,0 +1,360 @@
+/// qualify as such. Users may implement this trait for other types. E.g. it
+/// could be useful to have a `SmallVec` or a static buffer implementation for
+/// this trait.
+pub trait Accumulator {
+    /// Resets the buffer which cleans all state from it.
+    ///
+    /// # Note
+    ///
+    /// Useful when using `Vec` or similar as accumulator.
+    fn reset(&mut self);
+    /// Writes the given bytes into the buffer.
+    fn write(&mut self, bytes: &[u8]);
+    /// Returns a shared reference to the slice of the current state of the buffer.
+    fn as_slice(&self) -> &[u8];
+}
+
+impl Accumulator for Vec<u8> {
+    fn reset(&mut self) {
+        <Vec<_>>::clear(self)
+    }
+
+    fn write(&mut self, bytes: &[u8]) {
+        // This could theoretically be speed-up by using `unsafe` `set_len`
+        // and `[u8]` `copy_from_slice` methods.
+        <Vec<_>>::extend_from_slice(self, bytes)
+    }
+
+    fn as_slice(&self) -> &[u8] {
+        <Vec<_>>::as_slice(self)
+    }
+}
+
+impl<'a, T> Accumulator for &'a mut T
+where
+    T: Accumulator,
+{
+    fn reset(&mut self) {
+        <T as Accumulator>::reset(self)
+    }
+
+    fn write(&mut self, bytes: &[u8]) {
+        <T as Accumulator>::write(self, bytes)
+    }
+
+    fn as_slice(&self) -> &[u8] {
+        <T as Accumulator>::as_slice(self)
+    }
+}
+
+/// Wraps a bytes buffer and turns it into an accumulator.
+///
+/// # Panics
+///
+/// Upon hash calculation if the underlying buffer length does not suffice the
+/// needs of the accumulated hash buffer.
+pub struct Wrap<'a> {
+    /// The underlying wrapped buffer.
+    buffer: &'a mut [u8],
+    /// The current length of the filled area.
+    len: usize,
+}
+
+impl<'a> From<&'a mut [u8]> for Wrap<'a> {
+    fn from(buffer: &'a mut [u8]) -> Self {
+        Self { buffer, len: 0 }
+    }
+}
+
+impl<'a> Accumulator for Wrap<'a> {
+    fn reset(&mut self) {
+        self.len = 0;
+    }
+
+    fn write(&mut self, bytes: &[u8]) {
+        let len = self.len;
+        let bytes_len = bytes.len();
+        <[u8]>::copy_from_slice(&mut self.buffer[len..(len + bytes_len)], bytes);
+        self.len += bytes_len;
+    }
+
+    fn as_slice(&self) -> &[u8] {
+        &self.buffer[..self.len]
+    }
+}
+
+impl<T> Default for CryptoHasher<T, Vec<u8>>
+where
+    T: markers::Sealed,
+{
+    fn default() -> Self {
+        Self::from(Vec::new())
+    }
+}
+
+impl<T, A> From<A> for CryptoHasher<T, A>
+where
+    T: markers::Sealed,
+    A: Accumulator,
+{
+    fn from(mut accumulator: A) -> Self {
+        <A as Accumulator>::reset(&mut accumulator);
+        Self {
+            accumulator,
+            marker: PhantomData,
+        }
+    }
+}
+
+/// Hash functions that can put their results into an output buffer
+/// described by the `Output` type.
+///
+/// # Note
+///
+/// This is an internal trait used to deduplicate some implementations.
+pub trait FinishInto<Output> {
```

Comment:
Can `FinishInto` be implemented with multiple `Output` types for some definite type? I guess no. So it should be associated type for `FinishInto`, not a generic parameter?

Reply:
I can and it is: E.g. for `[u8; 32]`, `[u8; 16]` and `[u8; 8]` for the TwoX hasher.